src/pools/thread/dynamic.ts

```diff
@@ -47,31 +47,17 @@ export class DynamicThreadPool<
    * @returns Thread worker.
    */
   protected chooseWorker (): ThreadWorkerWithMessageChannel {
-    for (const [worker, numberOfTasks] of this.tasks) {
-      if (numberOfTasks === 0) {
-        // A worker is free, use it
-        return worker
-      }
-    }
-
-    if (this.workers.length === this.max) {
-      this.emitter.emit('FullPool')
-      return super.chooseWorker()
-    }
-
-    // All workers are busy, create a new worker
-    const workerCreated = this.createAndSetupWorker()
-    this.registerWorkerMessageListener<Data>(workerCreated, message => {
-      const tasksInProgress = this.tasks.get(workerCreated)
-      if (
-        isKillBehavior(KillBehaviors.HARD, message.kill) ||
-        tasksInProgress === 0
-      ) {
-        // Kill received from the worker, means that no new tasks are submitted to that worker for a while ( > maxInactiveTime)
-        this.sendToWorker(workerCreated, { kill: 1 })
-        void this.destroyWorker(workerCreated)
-      }
-    })
-    return workerCreated
+    return dynamicallyChooseWorker<ThreadWorkerWithMessageChannel, Data>(
+      this.tasks,
+      this.workers,
+      this.max,
+      this.emitter,
+      this.nextWorkerIndex,
+      nextIndex => (this.nextWorkerIndex = nextIndex),
+      this.createAndSetupWorker.bind(this),
```

Comment:
could we save the binded functions and reuse them instead of do the bind each time ? 
We could also try to draft a class instead of a single method so we can store those binded functions into the class instead of pass them each time ( performance reason ).

Please let me know your thoughts

Reply:
The current problem is, that most of these functions are protected by design.

Are there performance reasons?

So please provide / draft an alternative solution if you have one

Reply:
@Shinigami92  not sure I got your point could you please elaborate more on that?
We can just set the functions as protected you know, like we do with others

Yes there are performance reasons using the bind function each time will decrease the performance.

Reply:
Maybe I'm overlooking something, but when I pass e.g. the instance like `dynamicallyChooseWorker(this)` I don't have access to e.g. `that.createAndSetupWorker()` because `createAndSetupWorker` is a protected member function
So how should I pass these functions to `dynamicallyChooseWorker` so I can successfully call them within `dynamicallyChooseWorker`?

---

Feel free to create a PR stacked on this PR and show some experiments :slightly_smiling_face:

Reply:
`
// do this in some point where the code is executed one time ( maybe the constructor could be fine)
const myBindedFunction = this.createAndSetupWorker.bind(this)  

// or
this.myBindedFunction = this.createAndSetupWorker.bind(this) // in this case we could need to set it as protected in some way
`

and  


`
dynamicallyChooseWorker(..,..,..,myBindedFunction)
`